param.go

```diff
@@ -98,9 +98,36 @@ type paramVisitor interface {
 // recursed into.
 type paramVisitorFunc func(param) (recurse bool)
 
-func (f paramVisitorFunc) Visit(p param) paramVisitor {
-	if f(p) {
-		return f
+// paramVisitOnce encapsulates a paramVisitorFun and a map to track visited nodes.
+type paramVisitOnce struct {
+	f       paramVisitorFunc
+	visited map[string]bool
+}
+
+// NewParamVisitOnce constructs a paramVisitor avoiding redundant visits.
+func newParamVisitOnce(visited map[string]bool, f paramVisitorFunc) paramVisitor {
+	return paramVisitOnce{
+		f:       f,
+		visited: visited,
+	}
+}
+
+// Visit operates pv.f on param if:
+//		- param is not an edge node (i.e. param is a paramList), OR
+//		- param is an edge node that has not yet been visited
+func (pv paramVisitOnce) Visit(p param) paramVisitor {
+	switch p.(type) {
+	case paramSingle, paramGroupedSlice:
+		if pv.visited[p.String()] {
+			return pv
+		}
+		defer func() {
```

Comment:
This is an interesting though. Instead of the defer it appears to be just as valid to set the `visited` map before returning `pv` a few lines below. There is no real value is setting `visited` if we have already found a cycle (and therefore `pv.f()` is returning `false` to indicate no further recursion necessary.

As for the Annotation pattern from that comment, while I did not initially read that comment I still don't feel like this is the right approach. The utility of `paramVisitor` interface is in its usage by `walkParam`. The only argument or adding to this interface is if `walkParam` needs a new way to interact with the `paramVisitor` implementation. I would argue the encapsulation of the caching functionality inside the `paramVisitOnce` implementation while satisfying the `paramVisitor` interface is cleaner than providing explicit caching hooks. Maybe I am wrong though!

Reply:
I am convinced of the struct wrapper approach. I also tinkered with the defer and convinced myself that is the right choice, too.